File: benchmarks/bench.go
// Benchmark for gitverse.ru/pkg-ru/imager-client/v2 (Go implementation).
//
// Запуск из корня проекта:
//
//	go run docs/bench.go [iterations]
//
// Измеряет клиентские операции (формирование URL/AssetType/HTML) — чистые
// функции без HTTP. Админ-методы не бенчмаркастся.
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"

	imager "gitverse.ru/pkg-ru/imager-client/v2"
)

func bench(name string, fn func(), iterations int) {
	start := time.Now()
	for i := 0; i < iterations; i++ {
		fn()
	}
	elapsed := time.Since(start).Seconds()
	ops := float64(iterations) / elapsed
	us := (elapsed / float64(iterations)) * 1_000_000
	fmt.Printf("%s: %.0f ops/s, %.2f μs/op\n", name, ops, us)
}

func main() {
	im := imager.New(imager.Options{BaseURL: "https://cdn.example.com/"})

	n := 100_000
	if len(os.Args) > 1 {
		if parsed, err := strconv.Atoi(os.Args[1]); err == nil && parsed > 0 {
			n = parsed
		}
	}

	bench("GetAsset", func() { im.GetAsset("image.jpg", "200x200", "webp", nil) }, n)
	bench("GetAssets", func() { im.GetAssets("image.jpg", []any{"200x200", "400x400"}, []string{"webp", "avif"}, nil) }, n)
	bench("GetAssetPath", func() { im.GetAssetPath("image.jpg", "200x200", "webp", nil) }, n)
	bench("GetAssetsHtml", func() { im.GetAssetsHtml("image.jpg", []any{"200x200", "400x400"}, []string{"webp", "avif"}, nil, nil) }, n)
}

File: benchmarks/bench.ts
/**
 * Benchmark for imager-client TypeScript implementation.
 *
 * Запуск из корня проекта:
 *     npx tsx docs/bench.ts [iterations]
 *
 * Измеряет клиентские операции (формирование URL/AssetType/HTML) — чистые
 * функции без HTTP. Админ-методы не бенчмаркастся.
 */
import { Imager } from "../src/imager-ts/Imager";

const imager = new Imager({ baseURL: "https://cdn.example.com/" });

function bench(
    name: string,
    fn: () => void,
    iterations = 100_000,
): void {
    const start = performance.now();
    for (let i = 0; i < iterations; i++) fn();
    const elapsed = performance.now() - start;
    const ops = iterations / (elapsed / 1000);
    const us = (elapsed / iterations) * 1000;
    console.log(
        `${name}: ${ops.toLocaleString("en-US", { maximumFractionDigits: 0 })} ops/s, ${us.toFixed(2)} μs/op`,
    );
}

const n = process.argv[2] ? parseInt(process.argv[2], 10) : 100_000;

bench("GetAsset", () => imager.GetAsset("image.jpg", "200x200", "webp"), n);
bench(
    "GetAssets",
    () => imager.GetAssets("image.jpg", ["200x200", "400x400"], ["webp", "avif"]),
    n,
);
bench("GetAssetPath", () => imager.GetAssetPath("image.jpg", "200x200", "webp"), n);
bench(
    "GetAssetsHtml",
    () => imager.GetAssetsHtml("image.jpg", ["200x200", "400x400"], ["webp", "avif"]),
    n,
);
